runtime/Java/src/org/antlr/v4/runtime/tree/TreeRewriteSupport.java

```diff
@@ -0,0 +1,157 @@
+package org.antlr.v4.runtime.tree;
+
+
+import org.antlr.v4.runtime.ParserRuleContext;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/** This class encapsulates code necessary to support tree rewriting
+ *  using visitors. To rewrite trees using ANTLR-generated visitors,
+ *  override two key methods:
+ * <pre>
+	 public class MyTreeRewriter extends MyParserNameBaseVisitor<ParseTree> {
```

Comment:
💭 Currently you are using a return value of `null` to identify nodes that you want to remove. Another option would be not allowing node deletion in this way but instead require the parent be visited in order to remove one of its children. The advantage of this approach is you could avoid overriding `visitTerminal` and `visitErrorNode` every time, but it of course comes with the downside of making node removal potentially more difficult.

Roslyn's approach is more abstracted. Rather than provide the entire visitor as input, you only need to supply a transformation function (basically a `BiFunction<ParseTree, ParseTree, ParseTree>` for Java). The first argument to the transformation function is the original `ParseTree` that needs to be rewritten, and the second argument is the result of calling `rewriteChildren` (different from the first argument if and only if the children were modified). One could choose to implement this function with a visitor, or even something as simple as looking up values in a `Map<ParseTree, ParseTree>`.

Roslyn also provides a utility class similar to what we use `TokenStreamRewriter` for. This augments the rewriter functionality and simplifies it for certain uses, such as the following:

```java
ParseTree originalTree = ...;
TreeEditor editor = new TreeEditor(originalTree);
editor.replaceTerminal(someTerminal, newTerminal);
editor.replaceTree(someTree, newTree);
editor.remove(nodeToRemove);
ParseTree rewritten = editor.rewriteTree(); // Or perhaps editor.run()
```

:bulb: I would recommend at minimum adding such a `TreeEditor` to simplify many common cases for rewriting trees, so users don't have to worry about getting the visitor semantics exactly right.